pusher-platform-core/src/main/kotlin/com/pusher/platform/BaseClient.kt

```diff
@@ -36,120 +40,124 @@ data class BaseClient(
-            initialEventId = initialEventId,
+    ): Subscription = createResumingStrategy(
+        initialEventId = initialEventId,
+        logger = logger,
+        nextSubscribeStrategy = createTokenProvidingStrategy(
+            tokenProvider = tokenProvider,
+            tokenParams = tokenParams,
             logger = logger,
-            nextSubscribeStrategy = createTokenProvidingStrategy(
-                tokenProvider = tokenProvider,
-                tokenParams = tokenParams,
-                logger = logger,
-                nextSubscribeStrategy = createBaseSubscription(
-                    path = getRequestPath(requestDestination)
-                )
-            ),
-            errorResolver = ErrorResolver(connectivityHelper, retryOptions, scheduler)
-        )
-
-        return subscribeStrategy(listeners, headers)
-    }
-
-    fun subscribeNonResuming(
-        requestDestination: RequestDestination,
-        listeners: SubscriptionListeners,
-        headers: Headers,
+            nextSubscribeStrategy = createBaseSubscription<A>(
+                path = destination.toRequestPath(),
+                bodyParser = bodyParser
+            )
+        ),
+        errorResolver = ErrorResolver(connectivityHelper, retryOptions, scheduler)
+    )(listeners, headers)
+
+    fun <A> subscribeNonResuming(
+        destination: RequestDestination,
+        listeners: SubscriptionListeners<A>,
+        headers: ElementsHeaders,
         tokenProvider: TokenProvider?,
         tokenParams: Any?,
-        retryOptions: RetryStrategyOptions): Subscription {
-
-        val subscribeStrategy: SubscribeStrategy = createRetryingStrategy(
+        retryOptions: RetryStrategyOptions,
+        bodyParser: SubscriptionBodyParser<A>
+    ): Subscription = createRetryingStrategy(
+        logger = logger,
+        nextSubscribeStrategy = createTokenProvidingStrategy(
+            tokenProvider = tokenProvider,
+            tokenParams = tokenParams,
             logger = logger,
-            nextSubscribeStrategy = createTokenProvidingStrategy(
-                tokenProvider = tokenProvider,
-                tokenParams = tokenParams,
-                logger = logger,
-                nextSubscribeStrategy = createBaseSubscription(path = getRequestPath(requestDestination))),
-            errorResolver = ErrorResolver(connectivityHelper, retryOptions, scheduler)
-        )
-        return subscribeStrategy(listeners, headers)
-    }
+            nextSubscribeStrategy = createBaseSubscription<A>(path = destination.toRequestPath(), bodyParser = bodyParser)),
+        errorResolver = ErrorResolver(connectivityHelper, retryOptions, scheduler)
+    )(listeners, headers)
 
-    fun request(
+    @JvmOverloads
+    fun <A> request(
         requestDestination: RequestDestination,
-        headers: elements.Headers,
+        headers: ElementsHeaders,
         method: String,
+        responseType: Type,
```

Comment:
It's for the returned future. there is a chance that the wrong type is provided. The reason for using `Type` instead of `Class<A>` is because the later doesn't contain generic information but `Type` can be of this type: https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/ParameterizedType.html which will contain the extra information. With Gson this can be done using a `TypeToken` which we expose with the API as the inlined function `typeToken<A>()`. The only reason why Gson is still leaked as a dependency :(

An alternative I've thought is to provide the same `Parser` concept as with the subscriptions. I'll do that, as it makes it more typed and avoids having to expose Gson.

Reply:
I'm not sure I completely understand the change your suggested in terms of provide the same `Parser` concept as with subscriptions.